packages/react-compose/src/next/README.md

```diff
@@ -0,0 +1,226 @@
+# @fluentui/react-compose
+
+`@fluentui/react-compose` is a library of utilities for composing Fluent UI components.
+
+## A basic component walkthrough
+
+Building a recomposable component requires that we build legos; we put them together, but we let others put them together how they may need. Here's what's needed:
+
+- **State hooks** - all processing of the component should be done within one or more reusable hooks. The hooks should be allowed to manipulate the state in some reusable way, so that recomposing the component can be easily done.
+- **Render function** - a function which takes in final state and returns JSX. (e.g. `renderButton`)
+- **Component factory function** - a create function which returns `{ state, render }` parts for creating the component. This is used to scaffold a new version of the component as needed.
+- **Style hooks** - hooks which can take in styles and provide appropriate classnames to the anatomy of the component.
+
+With these building blocks, you can compose or recompose the component in numerous ways.
+
+### Simple example
+
+A render function:
+
+```jsx
+const renderButton = state => {
+  return <button {...state} />;
+};
+```
+
+A hook which can manipulate the state (defining accessibility and behaviors):
+
+```jsx
+const useButton = state => {
+  // If the button is rendered as something besides a button or anchor, make it focusable.
+  if (state.as !== 'button' && state.as !== 'a') {
+    state.tabIndex = 0;
+  }
+};
+```
+
+A factory function sets up the state and render function:
+
+```jsx
+const createButton = (userProps, ref, defaultProps) => {
+  const state = _.merge({}, defaultProps, userProps);
+
+  // Apply button behaviors.
+  useButton(state);
+
+  return { state, render };
+};
+```
+
+A button can now be easily scaffolded, along with your choice of styling systems:
+
+```jsx
+import * as styles from './Button.scss';
+
+const useStyles = makeStyles(styles);
+
+const Button = React.forwardRef((props, ref) => {
+  const { state, render } = createButton(props, ref);
+
+  // Inject classNames as needed.
+  useStyles(state);
+
+  return render(state);
+});
+```
+
+### Details
+
+#### Creating mutable state with `mergeProps`
+
+In the previous example, `_.merge` was used to deep clone the props into a state object. Creating a single clone and using that to construct state simplifies hook development and usage; rather than trying to re-clone objects unnecessarily on every small mutation, hooks can assume operating against a draft state. This creates more self contained hooks, which can ensure they are used correctly, avoiding accidents like stomping on existing event handlers by blind object assigning the results.
+
+However, deep merge overlooks many edge cases for component props:
+
+- Deep merging classnames should append them, not replace
+- Deep merging JSX, ref objects, or arrays should replace, not recurse/clone
+- Deep merging an object on a string should replace
```

Comment:
And how callbacks will be handled? I.e. `mergeProps({ onClick: props.onClick }, { onClick: () => {} })`?

Reply:
How many cases are you going to handle?

```js
const today = new Date();
const result = mergeProps({
  date: { as: "span", today }
});

today.setFullYear(2100);

console.log(today.toDateString(), result.date.today.toDateString()); // Tue Aug 03 2100 Tue Aug 03 2100 
```

Lodash handles this scenario.

Reply:
functions and arrays are pass through. it does not try to merge 2 callbacks together; it will replace.

we could also apply the deep merge only for Object prototypes and leave class instances as replace rather than merge.

I'm not following what you're referring to with lodash.

Reply:
I believe this is ok; but happy to discuss further. I've opened a related RFC here https://github.com/microsoft/fluentui/issues/14365 and we can go through some alternatives if you want tweaks or see a better direction!